mass2/core/channels.py

```diff
@@ -665,46 +667,49 @@ def save_analysis(self, path: Path | str, overwrite: bool = False) -> None:
         overwrite : bool, optional
             If `path` exists, whether to overwrite it, by default False
         """
-        if isinstance(path, str):
-            path = pathlib.Path(path)
-        if os.path.exists(path):
-            if os.path.exists(path / "data_all.pkl") and not overwrite:
-                raise ValueError(f"Set save_analysis(...overwrite=True) to overwrite the existing {path=}")
-            if not os.path.isdir(path):
-                raise ValueError(f"save_analysis(path={path}) requires that `path` be a directory, if it exists")
-        else:
-            os.mkdir(path)
-            assert os.path.exists(path) and os.path.isdir(path)
-
-        channels = {}
-        for ch_num, ch in self.channels.items():
-            fname = path / f"data_chan{ch_num:04d}.parquet"
-            df = ch.df.drop("pulse", "timestamp", "subframecount")
-            df.write_parquet(fname)
-            steps = ch.steps.trim_dead_ends(None, drop_debug=True)
-            channels[ch_num] = dataclasses.replace(ch, df=pl.DataFrame(), df_history=[], noise=None, steps=steps)
-        data = dataclasses.replace(self, channels=channels, bad_channels={})
-        pickle_file = path / "data_all.pkl"
-        pickle_object(data, pickle_file)
+        path = pathlib.Path(path)
+        if path.suffix != ".zip":
+            path = path.with_suffix(".zip")
+
+        if os.path.exists(path) and not overwrite:
+            raise ValueError(f"Set save_analysis(...overwrite=True) to overwrite the existing {path=}")
+        with ZipFile(str(path), "w") as zf:
+            channels = {}
+            for ch_num, ch in self.channels.items():
+                df = ch.df
+                columns_to_drop = {"pulse", "timestamp", "subframecount"}
+                columns_to_drop = columns_to_drop.intersection(df.columns)
+                for c in columns_to_drop:
+                    df = df.drop(c)
+                buffer = io.BytesIO()
+                df.write_parquet(buffer)
+                fname = f"data_chan{ch_num:04d}.parquet"
+                zf.writestr(fname, buffer.getvalue())
+                steps = ch.steps.trim_dead_ends(None, drop_debug=True)
+                channels[ch_num] = dataclasses.replace(ch, df=pl.DataFrame(), df_history=[], noise=None, steps=steps)
+            data = dataclasses.replace(self, channels=channels, bad_channels={})
+            pickle_file = "data_all.pkl"
+            bytes = pickle_stream(data)
```

Comment:
This may not matter too much.... but this will basically duplicate all data in memory while writing this file.

Reply:
Yeah. I couldn't figure out a better way, but I agree it does seem to copy everything to a string.